File: zest4096.c
#include <stdint.h>

// Rotate Left
#define ROTL(a, b) (((a) << (b)) | ((a) >> (64 - (b))))
// Perform an eighth of a round
#define RE(a, b, c, d, e, f, g, h) (	\
a += b, h = ROTL(h ^ a, 32), 			\
c += d, b = ROTL(b ^ c, 28), 			\
e += f, d = ROTL(d ^ e, 24), 			\
g += h, f = ROTL(f ^ g, 21),			\
a += b, h = ROTL(h ^ a, 16), 			\
c += d, b = ROTL(b ^ c, 12), 			\
e += f, d = ROTL(d ^ e,  8), 			\
g += h, f = ROTL(f ^ g,  5))

// Input and Output block size of 64 * 64, or 4096 bits (512 bytes)
void zest4096(uint64_t out[64], const uint64_t in[64]) {
	// Copy input to working value array
	uint64_t x[64];
	for (int i = 0; i < 64; ++i)
		x[i] = in[i];
	// Perform 48 * 2 rounds
	for (uint32_t i = 0; i < 48; i++) {
		// Odd round (columns)
		RE(x[0], x[ 8], x[16], x[24], x[32], x[40], x[48], x[56]);
		RE(x[1], x[ 9], x[17], x[25], x[33], x[41], x[49], x[57]);
		RE(x[2], x[10], x[18], x[26], x[34], x[42], x[50], x[58]);
		RE(x[3], x[11], x[19], x[27], x[35], x[43], x[51], x[59]);
		RE(x[4], x[12], x[20], x[28], x[36], x[44], x[52], x[60]);
		RE(x[5], x[13], x[21], x[29], x[37], x[45], x[53], x[61]);
		RE(x[6], x[14], x[22], x[30], x[38], x[46], x[54], x[62]);
		RE(x[7], x[15], x[23], x[31], x[39], x[47], x[55], x[63]);
		// Even round (diagonals)
		RE(x[0], x[ 9], x[18], x[27], x[36], x[45], x[54], x[63]);
		RE(x[1], x[10], x[19], x[28], x[37], x[46], x[55], x[56]);
		RE(x[2], x[11], x[20], x[29], x[38], x[47], x[48], x[57]);
		RE(x[3], x[12], x[21], x[30], x[39], x[40], x[49], x[58]);
		RE(x[4], x[13], x[22], x[31], x[32], x[41], x[50], x[59]);
		RE(x[5], x[14], x[23], x[24], x[33], x[42], x[51], x[60]);
		RE(x[6], x[15], x[16], x[25], x[34], x[43], x[52], x[61]);
		RE(x[7], x[ 8], x[17], x[26], x[35], x[44], x[53], x[62]);
	}
	// Add to original input
	for (int i = 0; i < 64; ++i)
		out[i] = x[i] + in[i];
}

File: zest4096.rs
use std::num::Wrapping;

#[inline(always)]
fn rotl(a: u64, b: u8) -> u64 {
    a << b | a >> (64 - b)
}

#[inline(always)]
fn eighth_round(e: (usize, usize, usize, usize, usize, usize, usize, usize), x: &mut [u64; 64]) {
    x[e.0] = (Wrapping(x[e.0]) + Wrapping(x[e.1])).0;
    x[e.7] = rotl(x[e.7] ^ x[e.0], 32);
    x[e.2] = (Wrapping(x[e.2]) + Wrapping(x[e.3])).0;
    x[e.1] = rotl(x[e.1] ^ x[e.2], 28);
    x[e.4] = (Wrapping(x[e.4]) + Wrapping(x[e.5])).0;
    x[e.3] = rotl(x[e.3] ^ x[e.4], 24);
    x[e.6] = (Wrapping(x[e.6]) + Wrapping(x[e.7])).0;
    x[e.5] = rotl(x[e.5] ^ x[e.6], 21);

    x[e.0] = (Wrapping(x[e.0]) + Wrapping(x[e.1])).0;
    x[e.7] = rotl(x[e.7] ^ x[e.0], 16);
    x[e.2] = (Wrapping(x[e.2]) + Wrapping(x[e.3])).0;
    x[e.1] = rotl(x[e.1] ^ x[e.2], 12);
    x[e.4] = (Wrapping(x[e.4]) + Wrapping(x[e.5])).0;
    x[e.3] = rotl(x[e.3] ^ x[e.4], 8);
    x[e.6] = (Wrapping(x[e.6]) + Wrapping(x[e.7])).0;
    x[e.5] = rotl(x[e.5] ^ x[e.6], 5);
}

pub fn zest_hash(input: [u64; 64]) -> [u64; 64] {
    let mut x = input;
    for _ in 0..48 {
        // Odd round (columns)
        eighth_round((0, 8, 16, 24, 32, 40, 48, 56), &mut x);
        eighth_round((1, 9, 17, 25, 33, 41, 49, 57), &mut x);
        eighth_round((2, 10, 18, 26, 34, 42, 50, 58), &mut x);
        eighth_round((3, 11, 19, 27, 35, 43, 51, 59), &mut x);
        eighth_round((4, 12, 20, 28, 36, 44, 52, 60), &mut x);
        eighth_round((5, 13, 21, 29, 37, 45, 53, 61), &mut x);
        eighth_round((6, 14, 22, 30, 38, 46, 54, 62), &mut x);
        eighth_round((7, 15, 23, 31, 39, 47, 55, 63), &mut x);
        // Even round (diagonals)
        eighth_round((0, 9, 18, 27, 36, 45, 54, 63), &mut x);
        eighth_round((1, 10, 19, 28, 37, 46, 55, 56), &mut x);
        eighth_round((2, 11, 20, 29, 38, 47, 48, 57), &mut x);
        eighth_round((3, 12, 21, 30, 39, 40, 49, 58), &mut x);
        eighth_round((4, 13, 22, 31, 32, 41, 50, 59), &mut x);
        eighth_round((5, 14, 23, 24, 33, 42, 51, 60), &mut x);
        eighth_round((6, 15, 16, 25, 34, 43, 52, 61), &mut x);
        eighth_round((7, 8, 17, 26, 35, 44, 53, 62), &mut x);
    }
    for (i, elem) in x.iter_mut().enumerate() {
        *elem += input[i]
    }
    x
}
